ffi/examples/read-table/schema.h

```diff
@@ -0,0 +1,259 @@
+  add_to_list(builder->lists+sibling_list_id, name_ptr, type);
+}
+
+void visit_string(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "string");
+}
+
+void visit_long(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "long");
+}
+
+void visit_integer(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "integer");
+}
+
+void visit_short(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "short");
+}
+
+void visit_byte(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "byte");
+}
+void visit_float(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "float");
+}
+
+void visit_double(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "double");
+}
+
+void visit_boolean(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "boolean");
+}
+
+void visit_binary(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "binary");
+}
+
+void visit_date(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "date");
+}
+
+void visit_timestamp(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "timestamp");
+}
+
+void visit_timestamp_ntz(void *data, uintptr_t sibling_list_id, struct KernelStringSlice name) {
+  visit_simple_type(data, sibling_list_id, name, "timestamp_ntz");
+}
+
+// free all the data in the builder (but not the builder itself, since that might be stack allocated)
+void free_builder(SchemaBuilder builder) {
+  for (int i = 0; i < builder.list_count; i++) {
+    SchemaItemList *list = (builder.lists)+i;
+    for (int j = 0; j < list->len; j++) {
+      SchemaItem *item = list->list+j;
+      free(item->name);
+      // don't free item->type, those are static strings
+      if (!strncmp(item->type, "decimal", 7)) {
+        // except decimal types, we malloc'd those :)
+        free(item->type);
+      }
+      // don't free item->children, it's just a list in the builder so will be freed by the outer
+      // loop.
+    }
+    free(list->list); // free all the items in this list (we alloc'd them together)
+  }
+  free(builder.lists);
+}
+
+// Print the schema of the snapshot
+void print_schema(const SnapshotHandle *snapshot) {
+  SchemaBuilder builder = {
+    .list_count = 0,
+    .lists = calloc(0, sizeof(SchemaItem*)),
+  };
+  EngineSchemaVisitor visitor = {
```

Comment:
Yeah, we would have to make a second copy in `ffi` just for passing over. 

It is annoying that it won't be a compile error, but at least we could have the kernel fail with a helpful error message, rather than some engine function being called with an enum value it doesn't understand which could do... anything.